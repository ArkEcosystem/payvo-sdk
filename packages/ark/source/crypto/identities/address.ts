import { Base58Check, Hash } from "@payvo/sdk-cryptography";

import { Network } from "../interfaces/networks";
import { MultiSignatureAsset, KeyPair } from "./contracts";
import { PublicKeyError } from "./errors";
import { getPubKeyHash } from "./helpers.js";
import { PublicKey } from "./public-key";

export class Address {
	public static fromPassphrase(passphrase: string, network?: Network): string {
		return Address.fromPublicKey(PublicKey.fromPassphrase(passphrase), network);
	}

	public static fromPublicKey(publicKey: string, network?: Network): string {
		if (!PublicKey.verify(publicKey)) {
			throw new PublicKeyError(publicKey);
		}

		const buffer: Buffer = Hash.ripemd160(Buffer.from(publicKey, "hex"));
		const payload: Buffer = Buffer.alloc(21);

		payload.writeUInt8(getPubKeyHash(network), 0);
		buffer.copy(payload, 1);

		return this.fromBuffer(payload);
	}

	public static fromWIF(wif: string, network?: Network): string {
		return Address.fromPublicKey(PublicKey.fromWIF(wif, network), network);
	}

	public static fromMultiSignatureAsset(asset: MultiSignatureAsset, network?: Network): string {
		return this.fromPublicKey(PublicKey.fromMultiSignatureAsset(asset), network);
	}

	public static fromPrivateKey(privateKey: KeyPair, network?: Network): string {
		return Address.fromPublicKey(privateKey.publicKey, network);
	}

	public static fromBuffer(buffer: Buffer): string {
		return Base58Check.encode(buffer);
	}

	public static toBuffer(
		address: string,
		network?: Network,
	): { addressBuffer: Buffer; addressError?: string } {
		const buffer: Buffer = Base58Check.decode(address);
		const result: { addressBuffer: Buffer; addressError?: string } = {
			addressBuffer: buffer,
		};

		if (buffer[0] !== getPubKeyHash(network)) {
			result.addressError = `Expected address network byte ${getPubKeyHash(network)}, but got ${buffer[0]}.`;
		}

		return result;
	}

	public static validate(address: string): boolean {
		try {
			return Base58Check.decode(address)[0] === getPubKeyHash();
		} catch (err) {
			return false;
		}
	}
}
