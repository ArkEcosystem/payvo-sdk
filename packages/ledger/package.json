{
	"name": "@payvo/sdk-ledger",
	"version": "4.0.0-next.20",
	"description": "Ledger for @payvo/sdk",
	"license": "MIT",
	"contributors": [],
	"type": "module",
	"exports": "./distribution/index.js",
	"types": "./distribution/index.d.ts",
	"files": [
		"/distribution"
	],
	"scripts": {
		"build": "pnpm run clean && tsc",
		"build:analyze": "webpack --config ../../webpack.analyze.cjs",
		"build:release": "webpack --config ../../webpack.config.cjs",
		"build:watch": "pnpm run clean && tsc -w",
		"clean": "rimraf .coverage distribution tmp"
	},
	"dependencies": {
		"@ledgerhq/hw-transport-u2f": "^5.36.0-deprecated",
		"@ledgerhq/hw-transport-webhid": "^6.19.0",
		"@ledgerhq/hw-transport-webusb": "^6.19.0",
		"@payvo/sdk": "workspace:*",
		"platform": "^1.3.6"
	},
	"devDependencies": {
		"@types/node": "^16.11.10"
	},
	"engines": {
		"node": ">=16.0.0"
	},
	"engineStrict": true
}
