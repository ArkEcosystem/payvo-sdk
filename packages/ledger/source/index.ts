// Based on https://github.com/near/near-ledger-js/blob/master/supportedTransports.js

import LedgerU2F from "@ledgerhq/hw-transport-u2f";
import LedgerHID from "@ledgerhq/hw-transport-webhid";
import LedgerUSB from "@ledgerhq/hw-transport-webusb";
import { Services } from "@payvo/sdk";
import platform from "platform";

export class LedgerTransportFactory {
	public static async create(): Promise<Services.LedgerTransport> {
		return new LedgerTransportFactory().#make();
	}

	async #make(): Promise<Services.LedgerTransport> {
		const [supportsHID, supportsUSB, supportsU2F] = await Promise.all([
			this.#supportsHID(),
			this.#supportsUSB(),
			this.#supportsU2F(),
		]);

		if (!supportsHID && !supportsUSB && !supportsU2F) {
			throw new Error("No transports appear to be supported.");
		}

		const supportedTransports: Services.LedgerTransport[] = [];

		if (supportsHID) {
			supportedTransports.push(LedgerHID);
		}

		if (supportsUSB) {
			supportedTransports.push(LedgerUSB);
		}

		if (supportsU2F) {
			supportedTransports.push(LedgerU2F);
		}

		for (const transport of supportedTransports) {
			try {
				return await transport.create();
			} catch (error) {
				if (error.name === "TransportOpenUserCancelled") {
					throw error;
				}
			}
		}

		throw new Error("Failed to connect through any transport.");
	}

	async #supportsHID(): Promise<boolean> {
		try {
			return await LedgerHID.isSupported();
		} catch {
			return false;
		}
	}

	async #supportsUSB(): Promise<boolean> {
		try {
			if (await LedgerUSB.isSupported()) {
				return platform.os.family !== "Windows" && platform.name !== "Opera";
			}

			return false;
		} catch {
			return false;
		}
	}

	async #supportsU2F(): Promise<boolean> {
		try {
			return await LedgerU2F.isSupported();
		} catch {
			return false;
		}
	}
}
