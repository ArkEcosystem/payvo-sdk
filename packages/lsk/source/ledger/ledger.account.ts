import { getPathArray } from "./ledger.helpers.js";

/**
 * Defines an Account to be used when communicating with ledger.
 */
export class LedgerAccount {
	#account = 0;
	#coinIndex = 134; // @TODO get slip44 from manifest

	/**
	 * Specify the account number.
	 * @param {number} newAccount
	 * @returns {this}
	 */
	public account(newAccount: number): this {
		this.#assertValidPath(newAccount);
		this.#account = newAccount;
		return this;
	}

	/**
	 * Derive the path using hardened entries.
	 * @returns {Buffer} defines the path in buffer form.
	 */
	public derivePath(): Buffer {
		const pathArray = getPathArray(`44'/${this.#coinIndex}'/${this.#account}'`);
		const buffer = Buffer.alloc(pathArray.length * 4);

		pathArray.forEach((r, index) => buffer.writeUInt32BE(r, index * 4));

		return buffer;
	}

	/**
	 * Asserts that the given param is a valid path (integer > 0)
	 */
	#assertValidPath(value: number) {
		if (!Number.isInteger(value)) {
			throw new TypeError('Param must be an integer');
		}

		if (value < 0) {
			throw new Error('Param must be greater than zero');
		}
	}
}
