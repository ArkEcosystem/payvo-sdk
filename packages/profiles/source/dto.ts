export * from "./signed-transaction.dto.js";
export * from "./transaction.collection";
export * from "./transaction.dto.js";
export * from "./transaction.mapper";
