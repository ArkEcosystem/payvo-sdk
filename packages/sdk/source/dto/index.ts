export * from "./confirmed-transaction.js";
export * from "./signed-transaction.contract.js";
export * from "./signed-transaction.js";
export * from "./unspent-transaction.js";
export * from "./wallet.js";
