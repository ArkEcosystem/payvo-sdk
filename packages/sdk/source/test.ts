/* istanbul ignore file */

import { randomUUID } from "crypto";
import { ConfigKey, ConfigRepository } from "./coins";
import { Container, BindingType } from "./ioc";
import { NetworkManifest } from "./networks";
import { BigNumberService } from "./services";

const createContainer = ({
	config,
	httpClient,
	manifest,
	meta,
}: {
	config?: ConfigRepository;
	httpClient: any;
	manifest: NetworkManifest;
	meta?: any;
}): Container => {
	config ??= new ConfigRepository({ network: manifest.id, httpClient });

	config.set(ConfigKey.Network, manifest);

	if (meta) {
		for (const [key, value] of Object.entries(meta)) {
			config.set(key, value);
		}
	}

	const container = new Container();
	container.constant(BindingType.ConfigRepository, config);
	container.constant(BindingType.HttpClient, httpClient);
	container.singleton(BindingType.BigNumberService, BigNumberService);

	return container;
};

export const createService = <T = any>({
	config,
	httpClient,
	manifest,
	meta,
	predicate,
	service,
}: {
	config?: ConfigRepository;
	httpClient: any;
	manifest: NetworkManifest;
	meta?: any;
	predicate: any;
	service: any;
}): T => {
	const container = createContainer({
		config,
		httpClient,
		manifest,
		meta,
	});

	if (predicate) {
		predicate(container);
	}

	return container.resolve(service);
};

export const createServiceAsync = async <T = any>({
	config,
	httpClient,
	manifest,
	meta,
	predicate,
	service,
}: {
	config?: ConfigRepository;
	httpClient: any;
	manifest: NetworkManifest;
	meta?: any;
	predicate: any;
	service: any;
}): Promise<T> => {
	const container = createContainer({
		config,
		httpClient,
		manifest,
		meta,
	});

	if (predicate) {
		await predicate(container);
	}

	const uuid: string = randomUUID();

	container.singleton(uuid, service);

	return container.get(uuid);
};
