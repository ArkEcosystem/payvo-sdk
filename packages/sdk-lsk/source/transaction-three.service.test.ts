import "jest-extended";

import { IoC, Services, Signatories } from "@payvo/sdk";

import { identity } from "../test/fixtures/identity";
import { createService } from "../test/mocking";
import { DataTransferObjects } from "./coin.dtos";
import { AddressService } from "./address.service";
import { ClientService } from "./client-three.service";
import { KeyPairService } from "./key-pair.service";
import { LedgerService } from "./ledger.service";
import { PublicKeyService } from "./public-key.service";
import { TransactionService } from "./transaction-three.service";
import { SignedTransactionData } from "./signed-transaction.dto";

let subject: TransactionService;

beforeAll(async () => {
	subject = createService(TransactionService, "lsk.testnet", (container) => {
		container.constant(IoC.BindingType.Container, container);
		container.singleton(IoC.BindingType.AddressService, AddressService);
		container.singleton(IoC.BindingType.ClientService, ClientService);
		container.constant(IoC.BindingType.DataTransferObjects, DataTransferObjects);
		container.singleton(IoC.BindingType.DataTransferObjectService, Services.AbstractDataTransferObjectService);
		container.singleton(IoC.BindingType.KeyPairService, KeyPairService);
		container.singleton(IoC.BindingType.LedgerService, LedgerService);
		container.singleton(IoC.BindingType.PublicKeyService, PublicKeyService);
	});
});

describe("TransactionService", () => {
	describe("#transfer", () => {
		it("should sign with an mnemonic", async () => {
			const result = await subject.transfer({
				signatory: new Signatories.Signatory(
					new Signatories.MnemonicSignatory({
						signingKey: identity.mnemonic,
						address: identity.address,
						publicKey: identity.publicKey,
						privateKey: identity.privateKey,
					}),
				),
				data: {
					amount: 1,
					to: identity.address,
				},
			});

			expect(result).toBeInstanceOf(SignedTransactionData);
		});

		// {"moduleID":2,"assetID":0,"senderPublicKey":"951ea6fd174ad81a20a3276f7f99bda35a28241d5760c9d45d1ae78b2f1a8dd1","nonce":"2n","fee":"207000n","signatures":["eebb923a6aa8ec4db630a708c4e269876e6a8440831f0b5f011cbf968089e5a189966028e031d7ce5ca2605ba80c35cc01deccee0826fd0e77d156b2cd013500"],"asset":{"recipientAddress":"fe4a135a02dd658903dd9db4963e91eb19194dc5","amount":"100000000n","data":""},"id":"892599e5fc288cc17fcd9a60e835437093b73349a905e63487a78e9439c50653"}

		// {"moduleID":2,"assetID":0,"senderPublicKey":"951ea6fd174ad81a20a3276f7f99bda35a28241d5760c9d45d1ae78b2f1a8dd1","nonce":"2n","fee":"207000n","signatures":["2faca873a0156e6d902d33e5f9b82a7f19b2d472786d1db7fac44ceb6dff8bc2d0b06b3ce307d0b157e13d50b9988fa7a2e6d36ef08154e181533f334bc2630d",""],"asset":{"recipientAddress":"fe4a135a02dd658903dd9db4963e91eb19194dc5","amount":"100000000n","data":""},"id":"7ed94d567b577a7aa483e25bed968a0adf0350bcd041866d62a97a363491b18e"}

		// {"moduleID":2,"assetID":0,"senderPublicKey":"951ea6fd174ad81a20a3276f7f99bda35a28241d5760c9d45d1ae78b2f1a8dd1","nonce":"2n","fee":"207000n","signatures":["2faca873a0156e6d902d33e5f9b82a7f19b2d472786d1db7fac44ceb6dff8bc2d0b06b3ce307d0b157e13d50b9988fa7a2e6d36ef08154e181533f334bc2630d","eebb923a6aa8ec4db630a708c4e269876e6a8440831f0b5f011cbf968089e5a189966028e031d7ce5ca2605ba80c35cc01deccee0826fd0e77d156b2cd013500"],"asset":{"recipientAddress":"fe4a135a02dd658903dd9db4963e91eb19194dc5","amount":"100000000n","data":""},"id":"3305b30c2b1664e559954f13f1581fb5e682aae35c6e6a7bb5c7b58c53a241b7"}

		it.only("should sign with a multi-signature", async () => {
			const wallet1 = {
				signingKey: "foil broccoli rare pony man umbrella visual cram wing rotate fall never",
				address: "lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p",
				publicKey: "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
			};

			const wallet2 = {
				signingKey: "penalty name learn right reason inherit peace mango guitar heart nature love",
				address: "lskn2de9mo9z3g9jvbpj4yjn84vrvjzcn5c5mon7a",
				publicKey: "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
			};

			const transaction1 = await subject.transfer({
				signatory: new Signatories.Signatory(
					new Signatories.MnemonicSignatory({
						signingKey: wallet1.signingKey,
						address: wallet1.address,
						publicKey: wallet1.publicKey,
						privateKey: identity.privateKey,
					}),
				),
				data: {
					amount: 1,
					to: wallet1.address,
				},
			});

			const transaction2 = await subject.multiSign(transaction1, {
				signatory: new Signatories.Signatory(
					new Signatories.MnemonicSignatory({
						signingKey: wallet2.signingKey,
						address: wallet2.address,
						publicKey: wallet2.publicKey,
						privateKey: identity.privateKey,
					}),
				),
			});

			console.log(transaction2);
		});
	});

	describe("#delegateRegistration", () => {
		it("should verify", async () => {
			const result = await subject.delegateRegistration({
				signatory: new Signatories.Signatory(
					new Signatories.MnemonicSignatory({
						signingKey: identity.mnemonic,
						address: identity.address,
						publicKey: identity.publicKey,
						privateKey: identity.privateKey,
					}),
				),
				data: {
					username: "johndoe",
				},
			});

			expect(result).toBeInstanceOf(SignedTransactionData);
		});
	});

	describe("#vote", () => {
		it("should verify", async () => {
			const result = await subject.vote({
				signatory: new Signatories.Signatory(
					new Signatories.MnemonicSignatory({
						signingKey: identity.mnemonic,
						address: identity.address,
						publicKey: identity.publicKey,
						privateKey: identity.privateKey,
					}),
				),
				data: {
					votes: [
						{
							id: identity.address,
							amount: 10,
						},
					],
					unvotes: [],
				},
			});

			expect(result).toBeInstanceOf(SignedTransactionData);
		});

		it("should fail to vote with a number that is not a multiple of 10", async () => {
			await expect(
				subject.vote({
					signatory: new Signatories.Signatory(
						new Signatories.MnemonicSignatory({
							signingKey: identity.mnemonic,
							address: identity.address,
							publicKey: identity.publicKey,
							privateKey: identity.privateKey,
						}),
					),
					data: {
						votes: [
							{
								id: identity.address,
								amount: 1,
							},
						],
						unvotes: [],
					},
				}),
			).rejects.toThrow(/not a multiple of 10/);
		});
	});

	describe.skip("#multiSign", () => {
		it("should verify", async () => {
			const wallet1 = {
				signingKey: "foil broccoli rare pony man umbrella visual cram wing rotate fall never",
				address: "lskp4agpmjwgw549xdrhgdt6dfwqrpvohgbkhyt8p",
				publicKey: "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
			};

			const wallet2 = {
				signingKey: "penalty name learn right reason inherit peace mango guitar heart nature love",
				address: "lskn2de9mo9z3g9jvbpj4yjn84vrvjzcn5c5mon7a",
				publicKey: "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
			};

			const transaction1 = await subject.multiSignature({
				signatory: new Signatories.Signatory(
					new Signatories.MnemonicSignatory({
						signingKey: wallet1.signingKey,
						address: wallet1.address,
						publicKey: wallet1.publicKey,
						privateKey: identity.privateKey,
					}),
				),
				data: {
					numberOfSignatures: 2,
					mandatoryKeys: [wallet1.publicKey, wallet2.publicKey],
					optionalKeys: [],
				},
			});

			expect(transaction1).toBeInstanceOf(SignedTransactionData);

			const transaction2 = await subject.multiSign(transaction1, {
				signatory: new Signatories.Signatory(
					new Signatories.MnemonicSignatory({
						signingKey: wallet2.signingKey,
						address: wallet2.address,
						publicKey: wallet2.publicKey,
						privateKey: identity.privateKey,
					}),
				),
			});

			expect(transaction2).toBeInstanceOf(SignedTransactionData);

			const transaction3 = await subject.multiSign(transaction2, {
				signatory: new Signatories.Signatory(
					new Signatories.MnemonicSignatory({
						signingKey: wallet1.signingKey,
						address: wallet1.address,
						publicKey: wallet1.publicKey,
						privateKey: identity.privateKey,
					}),
				),
			});

			expect(transaction3).toBeInstanceOf(SignedTransactionData);
		});
	});
});
