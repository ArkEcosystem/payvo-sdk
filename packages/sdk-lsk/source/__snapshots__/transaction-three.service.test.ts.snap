// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionService #multiSignature 1`] = `
SignedTransactionData {
  "bigNumberService": BigNumberService {
    "configRepository": ConfigRepository {},
  },
  "broadcastData": Object {
    "asset": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "assetID": 0,
    "fee": "314000",
    "id": "a218ca9189872bf8bf836f3e1ebfc7770752551ae4fc61b7bb45f87fee89171f",
    "moduleID": 4,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "300716832537cc8930a8608fbf2e8ccce869411935ed863f721617e3d8a73253e645380758cac47f1a20a115b68146512ac4d89aa1c7a499c34a10c39bc63001",
      "",
      "300716832537cc8930a8608fbf2e8ccce869411935ed863f721617e3d8a73253e645380758cac47f1a20a115b68146512ac4d89aa1c7a499c34a10c39bc63001",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
  "decimals": 8,
  "identifier": "a218ca9189872bf8bf836f3e1ebfc7770752551ae4fc61b7bb45f87fee89171f",
  "signedData": Object {
    "asset": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "assetID": 0,
    "fee": "314000",
    "id": "a218ca9189872bf8bf836f3e1ebfc7770752551ae4fc61b7bb45f87fee89171f",
    "moduleID": 4,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "300716832537cc8930a8608fbf2e8ccce869411935ed863f721617e3d8a73253e645380758cac47f1a20a115b68146512ac4d89aa1c7a499c34a10c39bc63001",
      "",
      "300716832537cc8930a8608fbf2e8ccce869411935ed863f721617e3d8a73253e645380758cac47f1a20a115b68146512ac4d89aa1c7a499c34a10c39bc63001",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
}
`;

exports[`TransactionService #multiSignature 2`] = `
SignedTransactionData {
  "bigNumberService": BigNumberService {
    "configRepository": ConfigRepository {},
  },
  "broadcastData": Object {
    "asset": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "assetID": 0,
    "fee": "314000",
    "id": "88647ef68cb91c5796843a571da994edf175876469f51c2a826afa7aa3dc2a34",
    "moduleID": 4,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "300716832537cc8930a8608fbf2e8ccce869411935ed863f721617e3d8a73253e645380758cac47f1a20a115b68146512ac4d89aa1c7a499c34a10c39bc63001",
      "e0fcd3edb757214bb5b21a34cba6e485b7d23262487b4f980bb30e833fb1c2a972fa425a7223366949389108527fab74bf1ac2a56d8b030b0ea0bbf87214f30b",
      "300716832537cc8930a8608fbf2e8ccce869411935ed863f721617e3d8a73253e645380758cac47f1a20a115b68146512ac4d89aa1c7a499c34a10c39bc63001",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
  "decimals": 8,
  "identifier": "88647ef68cb91c5796843a571da994edf175876469f51c2a826afa7aa3dc2a34",
  "signedData": Object {
    "asset": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "assetID": 0,
    "fee": "314000",
    "id": "88647ef68cb91c5796843a571da994edf175876469f51c2a826afa7aa3dc2a34",
    "moduleID": 4,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "300716832537cc8930a8608fbf2e8ccce869411935ed863f721617e3d8a73253e645380758cac47f1a20a115b68146512ac4d89aa1c7a499c34a10c39bc63001",
      "e0fcd3edb757214bb5b21a34cba6e485b7d23262487b4f980bb30e833fb1c2a972fa425a7223366949389108527fab74bf1ac2a56d8b030b0ea0bbf87214f30b",
      "300716832537cc8930a8608fbf2e8ccce869411935ed863f721617e3d8a73253e645380758cac47f1a20a115b68146512ac4d89aa1c7a499c34a10c39bc63001",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
}
`;

exports[`TransactionService #multiSignature 3`] = `
SignedTransactionData {
  "bigNumberService": BigNumberService {
    "configRepository": ConfigRepository {},
  },
  "broadcastData": Object {
    "asset": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "assetID": 0,
    "fee": "314000",
    "id": "808fe64c94e7e651342cf8102d25626c68024638e038aa298c6074feb23fe34a",
    "moduleID": 4,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "e8343b8618a2ea6babf9d9ceb1f5fd3aa8c79fd91b771c40ede0f5b49c32096be7eb2e114a4fb90e6f0f02c3a120834e712654fb14dd33c0374347ae4f3fa208",
      "e0fcd3edb757214bb5b21a34cba6e485b7d23262487b4f980bb30e833fb1c2a972fa425a7223366949389108527fab74bf1ac2a56d8b030b0ea0bbf87214f30b",
      "e8343b8618a2ea6babf9d9ceb1f5fd3aa8c79fd91b771c40ede0f5b49c32096be7eb2e114a4fb90e6f0f02c3a120834e712654fb14dd33c0374347ae4f3fa208",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
  "decimals": 8,
  "identifier": "808fe64c94e7e651342cf8102d25626c68024638e038aa298c6074feb23fe34a",
  "signedData": Object {
    "asset": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "assetID": 0,
    "fee": "314000",
    "id": "808fe64c94e7e651342cf8102d25626c68024638e038aa298c6074feb23fe34a",
    "moduleID": 4,
    "multiSignature": Object {
      "mandatoryKeys": Array [
        "5f7f98c50575a4a7e70a46ff35b72f4fe2a1ad3bc9a918b692d132d9c556bdf0",
        "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
      ],
      "numberOfSignatures": 2,
      "optionalKeys": Array [],
    },
    "nonce": "8",
    "senderPublicKey": "ac574896c846b59477a9115b952563938c48d0096b84846c0b634a621e1774ed",
    "signatures": Array [
      "e8343b8618a2ea6babf9d9ceb1f5fd3aa8c79fd91b771c40ede0f5b49c32096be7eb2e114a4fb90e6f0f02c3a120834e712654fb14dd33c0374347ae4f3fa208",
      "e0fcd3edb757214bb5b21a34cba6e485b7d23262487b4f980bb30e833fb1c2a972fa425a7223366949389108527fab74bf1ac2a56d8b030b0ea0bbf87214f30b",
      "e8343b8618a2ea6babf9d9ceb1f5fd3aa8c79fd91b771c40ede0f5b49c32096be7eb2e114a4fb90e6f0f02c3a120834e712654fb14dd33c0374347ae4f3fa208",
    ],
    "timestamp": "2021-01-01T12:00:00.000Z",
  },
}
`;
